File: 01-Programming-Languages/Go/Data-Structures/HashTable/hash_map.go
/*
Problem: Hash Map Implementation
Implement a simple hash map with put, get, and delete operations using separate chaining.

Approach:
- Use an array of buckets, each bucket is a linked list of key-value pairs
- Compute bucket index using hash function (modulo)
- Handle collisions with separate chaining

Time Complexity: O(1) average, O(n) worst case
Space Complexity: O(n)

Example:
Input: put("key1", "value1"), get("key1"), delete("key1"), get("key1")
Output: "value1", ""
*/

package main

import "fmt"

type entry struct {
	key   string
	value string
	next  *entry
}

type HashMap struct {
	buckets []*entry
	size    int
}

func NewHashMap(capacity int) *HashMap {
	return &HashMap{buckets: make([]*entry, capacity), size: capacity}
}

func (m *HashMap) hash(key string) int {
	h := 0
	for _, ch := range key {
		h = h*31 + int(ch)
	}
	if h < 0 {
		h = -h
	}
	return h % m.size
}

func (m *HashMap) Put(key, value string) {
	idx := m.hash(key)
	e := m.buckets[idx]
	for e != nil {
		if e.key == key {
			e.value = value
			return
		}
		e = e.next
	}
	m.buckets[idx] = &entry{key: key, value: value, next: m.buckets[idx]}
}

func (m *HashMap) Get(key string) string {
	idx := m.hash(key)
	e := m.buckets[idx]
	for e != nil {
		if e.key == key {
			return e.value
		}
		e = e.next
	}
	return ""
}

func (m *HashMap) Delete(key string) {
	idx := m.hash(key)
	e := m.buckets[idx]
	var prev *entry
	for e != nil {
		if e.key == key {
			if prev == nil {
				m.buckets[idx] = e.next
			} else {
				prev.next = e.next
			}
			return
		}
		prev = e
		e = e.next
	}
}

func main() {
	hm := NewHashMap(16)
	hm.Put("key1", "value1")
	fmt.Println("Get key1:", hm.Get("key1"))
	hm.Delete("key1")
	fmt.Println("Get key1 after delete:", hm.Get("key1"))
}

File: 01-Programming-Languages/Rust/Data-Structures/HashTable/hash_map.rs
/*
Problem: Hash Map Implementation
Implement a simple hash map with put, get, and delete operations using separate chaining.

Approach:
- Use a vector of buckets, each bucket is a linked list of key-value pairs
- Compute bucket index using hash function (modulo)
- Handle collisions with separate chaining

Time Complexity: O(1) average, O(n) worst case
Space Complexity: O(n)

Example:
Input: put("key1", "value1"), get("key1"), delete("key1"), get("key1")
Output: Some("value1"), None
*/

struct Entry {
    key: String,
    value: String,
    next: Option<Box<Entry>>,
}

struct HashMap {
    buckets: Vec<Option<Box<Entry>>>,
    size: usize,
}

impl HashMap {
    fn new(size: usize) -> Self {
        let mut buckets = Vec::with_capacity(size);
        for _ in 0..size {
            buckets.push(None);
        }
        HashMap { buckets, size }
    }

    fn hash(&self, key: &str) -> usize {
        let mut h: usize = 0;
        for ch in key.chars() {
            h = h.wrapping_mul(31).wrapping_add(ch as usize);
        }
        h % self.size
    }

    fn put(&mut self, key: &str, value: &str) {
        let idx = self.hash(key);
        let mut curr = self.buckets[idx].as_mut();
        while let Some(entry) = curr {
            if entry.key == key {
                entry.value = value.to_string();
                return;
            }
            curr = entry.next.as_mut();
        }
        self.buckets[idx] = Some(Box::new(Entry {
            key: key.to_string(),
            value: value.to_string(),
            next: self.buckets[idx].take(),
        }));
    }

    fn get(&self, key: &str) -> Option<&str> {
        let idx = self.hash(key);
        let mut curr = self.buckets[idx].as_ref();
        while let Some(entry) = curr {
            if entry.key == key {
                return Some(&entry.value);
            }
            curr = entry.next.as_ref();
        }
        None
    }

    fn delete(&mut self, key: &str) {
        let idx = self.hash(key);
        let mut curr = &mut self.buckets[idx];
        loop {
            match curr {
                None => return,
                Some(entry) if entry.key == key => {
                    *curr = entry.next.take();
                    return;
                }
                Some(entry) => {
                    curr = &mut entry.next;
                }
            }
        }
    }
}

fn main() {
    let mut hm = HashMap::new(16);
    hm.put("key1", "value1");
    println!("Get key1: {:?}", hm.get("key1"));
    hm.delete("key1");
    println!("Get key1 after delete: {:?}", hm.get("key1"));
}
